RecoLocalTracker/Phase2TrackerRecHits/plugins/alpaka/Phase2OTRecHitsSoAConverter.cc

```diff
@@ -0,0 +1,265 @@
+
+    // Count the number of P hits in the OT to dimension the SoA
+    int PHitsInOTBarrel = 0;
+    for (const auto& detSet : stripHits) {
+      for (const auto& recHit : detSet) {
+        DetId detId(recHit.geographicalId());
+        if (detIdIsP_[detId.rawId()])
+          PHitsInOTBarrel++;
+      }
+    }
+    LogDebug("Phase2OTRecHitsSoAConverter")
+        << "Tot number of modules in Pixels " << modulesInPixel_ << '\n'
+        << "Tot number of p_modulesInPSInOTBarrel: " << orderedModules_.size() << '\n'
+        << "Number of strip (active) modules:      " << activeStripModules << '\n'
+        << "Number of strip hits: " << nStripHits << '\n'
+        << "Total hits of PinOTBarrel:   " << PHitsInOTBarrel << '\n';
+
+    HitsHost stripHitsHost(queue, PHitsInOTBarrel, orderedModules_.size());
+    auto& stripHitsModuleView = stripHitsHost.view<::reco::HitModuleSoA>();
+
+    std::vector<int> counterOfHitsPerModule(orderedModules_.size(), 0);
+    assert(!orderedModules_.empty());
+    for (const auto& detSet : stripHits) {
+      auto firstHit = detSet.begin();
+      auto detId = firstHit->rawId();
+      auto index = detIdToIndex_[detId];
+      int offset = 0;
+      if (detIdIsP_[detId]) {
+        offset = moduleIndexToOffset_[index];
+        counterOfHitsPerModule[offset] = detSet.size();
+      }
+    }
+#ifdef HITS_DEBUG
+    int modId = 0;
+    for (auto c : counterOfHitsPerModule) {
+      std::cout << "On module " << modId << " we have " << c << " hits." << std::endl;
+      modId++;
+    }
+#endif
+
+    std::vector<int> cumulativeHitPerModule(counterOfHitsPerModule.size());
+    std::partial_sum(counterOfHitsPerModule.begin(), counterOfHitsPerModule.end(), cumulativeHitPerModule.begin());
+    // Create new vector with first element as 0, then shifted contents from counterOfHitsPerModule
+    std::vector<int> shifted(cumulativeHitPerModule.size(), 0);
+    stripHitsModuleView[0].moduleStart() = nPixelHits;
+    LogDebug("Phase2OTRecHitsSoAConverter")
+        << "Module start: 0 with hits: " << stripHitsModuleView[0].moduleStart() << '\n';
+    for (size_t i = 1; i < cumulativeHitPerModule.size(); ++i) {
+      shifted[i] = cumulativeHitPerModule[i - 1];
+      stripHitsModuleView[i].moduleStart() = cumulativeHitPerModule[i - 1] + nPixelHits;
+      LogDebug("Phase2OTRecHitsSoAConverter")
+          << "Module start: " << i << " with hits: " << stripHitsModuleView[i].moduleStart() << '\n';
+    }
+
+    for (const auto& detSet : stripHits) {
+      auto firstHit = detSet.begin();
+      auto detId = firstHit->rawId();
+      auto det = trackerGeometry->idToDet(detId);
+      auto index = detIdToIndex_[detId];
+      int offset = 0;
+      if (detIdIsP_[detId]) {
+        offset = moduleIndexToOffset_[index];
+        for (const auto& recHit : detSet) {
+          // Select only P-hits from the OT barrel
+          if (detIdIsP_[detId]) {
+            int idx = shifted[offset]++;
```

Comment:
From a very quick glance I think you could use a local variable instead of updating `shifted` ?
The `shifted` could be removed altogether, using `cumulativeHitPerModule` directly ?